Update logs stream (#3113)
elastic/observability-docs: docs/en/observability/logs-stream.asciidoc

[[logs-stream]]
= Stream a log file

In this guide, you'll learn how to send a log file to Elasticsearch using a standalone {agent}, configure the {agent} and your data streams using the `elastic-agent.yml` file, and query your logs using the data streams you've set up. 

[discrete]
[[logs-stream-prereq]]
== Prerequisites

include::logs-metrics-get-started.asciidoc[tag=monitoring-prereqs]

[discrete]
[[logs-stream-install-config-agent]]
== Install and configure the standalone {agent}

Complete the following steps to install and configure the standalone {agent} and send your log data to {es}:

. <<logs-stream-extract-agent, Download and extract the {agent} installation package.>>
. <<logs-stream-install-agent, Install and start the {agent}.>>
. <<logs-stream-agent-config, Configure the {agent}.>>

[discrete]
[[logs-stream-extract-agent]]
=== Step 1: Download and extract the {agent} installation package

On your host, download and extract the installation package that corresponds with your system:

include::{ingest-docs-root}/docs/en/ingest-management/tab-widgets/download-widget.asciidoc[]

[discrete]
[[logs-stream-install-agent]]
=== Step 2: Install and start the {agent}

> After downloading and extracting the installation package, you're ready to install the {agent}. From the agent directory, run the the install command that corresponds with your system:

Reviewer: ```suggestion
After downloading and extracting the installation package, you're ready to install the {agent}. From the agent directory, run the install command that corresponds with your system:
```

> NOTE: On macOS, Linux (tar package), and Windows, run the `install` command to
> install {agent} as a managed service and start the service. The DEB and RPM
> packages include a service unit for Linux systems with
> systemd, so just enable then start the service.

Reviewer: I realize this was copied from the Fleet Guide, but this text is confusing to me.

> On macOS, Linux (tar package), and Windows, run the install command to install Elastic Agent as a managed service and start the service.

This sounds like you have to install and then start, but I think you only have to install, right?

> The DEB and RPM packages include a service unit for Linux systems with systemd, so just enable then start the service.

This sounds like you have to install and then start, which you do. Right?

What about this instead?

```suggestion
NOTE: On macOS, Linux (tar package), and Windows, run the `install` command to
install and start {agent} as a managed service. The DEB and RPM
packages include a service unit for Linux systems with
systemd. For these systems, you must enable and start the service.
```

If you do decide to update this snippet, please update the source file too: https://github.com/elastic/ingest-docs/blame/main/docs/en/ingest-management/elastic-agent/install-standalone-elastic-agent.asciidoc

include::{ingest-docs-root}/docs/en/ingest-management/tab-widgets/run-standalone-widget.asciidoc[]

During installation, you're prompted with some questions:

. When asked if you want to install the agent as a service, enter `Y`. 
. When asked if you want to enroll the agent in Fleet, enter `n`.

[discrete]
[[logs-stream-agent-config]]
=== Step 3: Configure the {agent}

With your agent installed, configure it by updating the `elastic-agent.yml` file. 

[discrete]
[[logs-stream-yml-location]]
==== Locate your configuration file

> After installing the agent, you'll find the `elastic-agent.yml` in the one of the following locations according to your system:

Reviewer: ```suggestion
After installing the agent, you'll find the `elastic-agent.yml` in one of the following locations according to your system:
```

[discrete]
[[logs-stream-example-config]]
==== Update your configuration file

The following is an example of a standalone {agent} configuration. To configure your {agent}, replace the contents of the `elastic-agent.yml` file with this configuration:

[source,yaml]
----
outputs:
  default:
    type: elasticsearch
    hosts: '<your-elasticsearch-endpoint>:<port>'
    api_key: 'your-api-key'
inputs:
  - id: your-log-id
    type: filestream
    streams:
      - id: your-log-stream-id
        data_stream:
          dataset: generic
        paths:
          - /var/log/your-logs.log
----

Next, set the values for the following fields:

. `hosts` – Copy the {es} endpoint from your deployment's page and add the port (the default port is `443`). For example, `https://my-deployment.es.us-central1.gcp.cloud.es.io:443`.
+
--
[role="screenshot"]
image::images/es-endpoint-cluster-id.png[{es} endpoint and cluster id location, 50%]
--
. `api-key` – Use an API key to grant the agent access to {es}. To create an API key for your agent, see {fleet-guide}/grant-access-to-elasticsearch.html#create-api-key-standalone-agent[Create API keys for standalone agents].
+

> NOTE: The API key format should be <id>:<key>. Make sure you selected *Beats* when you created your API key. Base64 encoded API keys are not currently supported in this configuration.

Reviewer: ```suggestion
NOTE: The API key format should be `<id>:<key>`. Make sure you selected *Beats* when you created your API key. Base64 encoded API keys are not currently supported in this configuration.
```

. `inputs.id` – A unique identifier for your input.

> . `type` – The type of input. For collectings logs, set this to `filestream`.

Reviewer: ```suggestion
. `type` – The type of input. For collecting logs, set this to `filestream`.
```

[[logs-stream]]
= Stream a log file

> In this guide, you'll learn how to take a log file and send it to Elasticsearch using a standalone {agent}. You'll configure the {agent} and your data streams using the Custom Logs integration. From there, you'll learn how to query your logs and use the data streams you've set up to have more control and flexibility when filtering your log data. 

Reviewer 2: You mention the "Custom logs integration" here, but that is not the case anymore as it was removed from the flow.

include::{ingest-docs-root}/docs/en/ingest-management/tab-widgets/download-widget.asciidoc[]

[discrete]
[[logs-stream-install-agent]]
== Step 2: Install and start the {agent}

> After downloading and extracting the installation package, you're ready to install the {agent}. From the agent directory, run the the command that corresponds with your system to install the {agent} as a service. Do *not* enroll the agent in Fleet.

Reviewer 2: You mention here "Do not enroll ...". As Fleet and enroll is not part of the guide, is there a chance the user would do this? If not, this introduces new terms the user does not even know about so would rather leave it out.

Author: You're prompted to enroll in Fleet when installing, so they will be introduced to it there.  I can clarify to say "Do not  enroll the agent in Fleet when prompted during installation" or ignore it if you think a user might be familiar enough with the concepts of Fleet and standalone agents at this point.

Reviewer 2: I doubt a user is familiar is enough and shouldn't be. ++ on having it and potentially get this step out of the way.

NOTE: On macOS, Linux (tar package), and Windows, run the `install` command to
install {agent} as a managed service and start the service. The DEB and RPM
packages include a service unit for Linux systems with
systemd, so just enable then start the service.

include::{ingest-docs-root}/docs/en/ingest-management/tab-widgets/run-standalone-widget.asciidoc[]

[discrete]
[[logs-stream-agent-config]]
== Step 3: Configure the {agent}

With your agent installed, you can configure it by updating the 'elastic-agent.yml' in the directory described in the {fleet-guide}/installation-layout.html[installation layout] that matches your system.

The following is an example of a standalone {agent} configuration:

[source,yaml]
----
outputs:
  default:
    type: elasticsearch
    hosts: '<your-elasticsearch-url>:<port>'
    api_key: 'your-api-key'
inputs:
  - id: your-log-id
    type: filestream
    streams:
      - id: your-log-stream-id
        data_stream:
          dataset: generic
        paths:
          - /var/log/your-logs.log
----

To configure your {agent}, set the values for the fields in the example configuration:

. `hosts` – Your {es} URL and the port. On {ecloud}, this . 

> .. Copy the {es} endpoint and cluster ID and from your deployment's page.

Reviewer 2: As discussed, only the endpoint should be needed. This will also simplify the text below.

Author: When I tested with just the endpoint, I wasn't able to ingest any logs data. I also tried with/without the port, and I was only able to ingest data with the URL in this format. Maybe I'm doing something wrong there?

Author: I tested this again, and was able to get it working with the endpoint URL if I had the port added, so we can simplify at least that far.

Author: @ruflin I used 443. I just tested again on a fresh deployment and got the same result. Is it possible I don't have something set up correctly that would require manually entering the port?

+
--
[role="screenshot"]
image::images/es-endpoint-cluster-id.png[{es} endpoint and cluster id location, 50%]
--
+
.. Your host URL is formatted as `https://CLUSTER_ID.REGION.CLOUD_PLATFORM.DOMAIN:PORT`. The default port is 443. For example, `https://ec47fc4d2c53414e1307e85726d4b9bb.us-east-1.aws.found.io:443`.
. `api-key` – An API key to grant the agent access to {es}. To create an API key for your agent, see {fleet-guide}/grant-access-to-elasticsearch.html#create-api-key-standalone-agent[Create API keys for standalone agents].
+
NOTE: The API key format should be <id>:<key>. Make sure you selected *Beats* when you created your API key. Base64 encoded API keys are not currently supported in this configuration.
. `inputs.id` – A unique identifier for your input.
. `type` – The type of input. For collectings logs, set this to `filestream`.
. `streams.id` – A unique identifier for your stream of log data. 
. `data_stream.dataset` – The name for your dataset data stream. You can name this data stream anything that signifies the source of the data. The default value is `generic`.

> . `paths` – The path to your log files. You can use a pattern like `/var/log/foo.log*`.

Reviewer 2: @mdbirnstiehl After the changes were made the config, I think the user has to restart elastic-agent or will it pick up all the changes directly?

Author: In the testing I've done, it picks it up without restarting.

Reviewer 2: My assumption is that it picks up changes around the path but not around the output. But I also have seen during my testing that sometimes I felt like it did not pick up changes / not full consistency but could not 100% reproduce it. @cmacknz Does this ring a bell?

[[logs-stream-agent-config]]
== Step 2: Configure the {agent}

Configure the {agent} by updating the 'elastic-agent.yml' in the folder where you extracted the installation package.

The following is an example of a simple standalone {agent} configuration:

[source,yaml]
----
outputs:
  default:
    type: elasticsearch
    hosts: '<your-elasticsearch-url>:<port>'
    api_key: 'your-api-key'
inputs:
  - id: your-log-id
    type: logfile

>     data_stream:

Reviewer 2: It is possible to just leave out the namespace and use default. I wonder if we should skip it at this stage?

      namespace: default
    streams:
      - id: your-log-stream-id
        data_stream:
          dataset: generic
        paths:
          - /var/log/your-logs.log
----

To configure your {agent}, set the values for the fields in the example configuration:

. `hosts` – Your {es} URL and the port.

> . `api-key` – An API key to grant the agent access to {es}. To create an API key for your agent, see {fleet-guide}/grant-access-to-elasticsearch.html#create-api-key-standalone-agent[Create API keys for standalone agents].

Reviewer 2: The link to the API key creation is ok for now. But I think this could become a stumbling point. We should improve this on the UX side ( @grabowskit ). Ideally we could give the user something they can just copy paste into Dev Tools or Click a single button to the get API key (we have this planned) instead of having to go to the right place, put the right json in and then select the encoding.

+

> NOTE: The API key format should be <id>:<key>. Make sure you selected *Beats* when you created your API key. Base64 encoded API keys are not currently supported in this configuration.

Reviewer 3: No, no idea why it is like this. That was before my time.

> NOTE: The API key format should be <id>:<key>. Make sure you selected *Beats* when you created your API key. Base64 encoded API keys are not currently supported in this configuration.

Reviewer 2: @ogupte Have you stumbled over this during creation of the onboarding script? Anything special you do to the API key?

> NOTE: The API key format should be <id>:<key>. Make sure you selected *Beats* when you created your API key. Base64 encoded API keys are not currently supported in this configuration.

Reviewer: Unfortunately, Oliver no longer works at Elastic.

> NOTE: The API key format should be <id>:<key>. Make sure you selected *Beats* when you created your API key. Base64 encoded API keys are not currently supported in this configuration.

Reviewer: @dedemorton do you remember the history here?

. `inputs.id` – A unique identifier for your input.

> . `type` – The type of input. For collectings logs, set this to `logfile`.

Reviewer 2: @belimawr I assume we should jump here to filestream?

Reviewer 4: Yes!

. `data_stream.namespace` – The name for your namespace data stream. The namespace is useful for grouping data such as an environment (dev, prod, or qa), a team, or a strategic business unit. Using a namespace makes it easier to search for data from a given source by using a matching pattern. The default value is `default`.
. `streams.id` – A unique identifier for your stream of log data. 
. `data_stream.dataset` – the name for your dataset data stream. You can name this data stream anything that signifies the source of the data. The default value is `generic`.

> . `paths` – the path to your log file.

Reviewer 2: Lets mention here that it can be a pattern like /var/log/foo.log*

[discrete]

> [[logs-stream-install-agent]]

Reviewer 2: The user should first install the elastic agent and then configure it?

image::images/stream-logs-example.png[example search query on the logs stream page in {kib}]
--

[discrete]
[[logs-stream-troubleshooting]]
= Troubleshooting 

If you're not seeing your log files in {kib}, check the following in the `elastic-agent.yml` file:

> NOTE: After installing the {agent}, you need to update the `elastic-agent.yml` in the directory described in the {fleet-guide}/installation-layout.html[installation layout].

Reviewer 2: See above, the user should directly modify the elastic-agent.yml after installation.

- Verify that the path to your logs file under `paths` is correct.

> - Verify that your API key is in <id>:<key> format. If not, your API key may be in an unsupported format, and you'll need to create an API key in *Beats* format. 

Reviewer 2: There are commands like `elastic-agent status` that we could recommend here for debugging. @belimawr Any recommendations or links we should point to?

Reviewer 4: The first things that come to mind are:
- [Troubleshooting guide](https://www.elastic.co/guide/en/fleet/current/fleet-troubleshooting.html)
